[SOURCE] Mailchimp: New Event Sources

sergioeliot2039 opened the pull request:
A new PR for Mailchimp sources. There existed a previous [one](https://github.com/PipedreamHQ/pipedream/pull/1239) that unintendedly had modifications to Reddit files. This time only file related to Mailchimp are included.

Here are some notes:

New Segment Subscribers, don't have a since and before parameters to narrow down pagination, so currently they paginate across all items.
New List from the related Pipedream Issue for Mailchimp sources is the same as New Audience (there was a duplicated source in the issue)
New or Updated Subscriber was implemented as New Subscriber, and Email Address change or a Profile change (3 different events to Mailchimp)
New List Events is a superset to New or Updated Subscriber because it exposes all available events related to Audience Lists. However, I thought having New or Updated Subscriber as a different source could be relevant to Mailchimp users.

jverce commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,129 @@
+    let beforeCreatedAt;
+    let sinceUpdatedAt;
+    let beforeUpdatedAt;
+    if (["Created"].includes(this.watchFor)) {
```
I've seen this pattern before in the PR.
Why are you comparing strings like this instead of `this.watchFor === "Created"`?

sergioeliot2039 replied:
Oh, I happened to use the Array's object includes method to check if an array item equals some string, then I somehow continued using it for regular string comparison.
I have adjusted the code base to use the string equal comparer.

jverce commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,129 @@
+      offset = offset + mailchimpAudienceSegments.length;
+    } while (mailchimpAudienceSegments.length > 0);
+  },
+};
```
We should refactor this to reuse the logic suggested in `new-campaign` (in terms of "sent" vs "created")

sergioeliot2039 replied:
This has been addressed.

jverce commented on components/mailchimp/sources/new-or-updated-subscriber/new-or-updated-subscriber.js:
```diff
@@ -0,0 +1,50 @@
+      return ["subscribe", "profile", "upemail"];
+    },
+    getEventType() {
+      return ["subscribe", "profile", "upemail"];
```
No `unsubscribe` or `cleaned`?

sergioeliot2039 replied:
Unsure why were not included. Zapier has that same trigger but it doesn't specify which evets are supported.
I have included `unsubscribe `or cleaned as events to watch for when subscriber is updated.

jverce commented on components/mailchimp/sources/new-or-updated-subscriber/new-or-updated-subscriber.js:
```diff
@@ -0,0 +1,50 @@
+    generateMeta(eventPayload) {
+      const ts = +new Date(eventPayload.fired_at);
+      let summary;
+      if (["subscribe"].includes(eventPayload.type)) {
+        summary = `${eventPayload["data[email]"]} subscribed to list ${eventPayload["data[list_id]"]}`;
+      } else if (["profile"].includes(eventPayload.type)) {
+        summary = `${eventPayload["data[email]"]}'s profile was updated`;
+      } else if (["upemail"].includes(eventPayload.type)) {
+        summary = `${eventPayload["data[old_email]"]}'s email address changed to ${eventPayload["data[new_email]"]}`;
+      } else {
+        console.log("Unexpected trigger found, skipping emit");
+        return;
+      }
```
No need to make summaries too fancy. 
Something like the suggestion in `new-list-events` is enough.

sergioeliot2039 replied:
This has been addressed.

jverce commented on components/mailchimp/sources/new-or-updated-subscriber/new-or-updated-subscriber.js:
```diff
@@ -0,0 +1,50 @@
+        return;
+      }
+      return {
+        id: `${eventPayload["data[id]"]}`,
```
No need to use a template string here
```suggestion
        id: eventPayload["data[id]"],
```

jverce commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,115 @@
+        "Watch only for new orders with a specific Campaign Id value.",
+      optional: true,
+    },
+    outreachId: {
+      type: "string",
+      label: "Outreach Id",
+      description:
+        "Watch only for new orders with a specific Outreach Id value.",
+      optional: true,
+    },
```
This one should go after `hasOutreach`, and contain a message saying "ignored if `hasOutreach` is marked as 'No'"

sergioeliot2039 replied:
This was changed.

jverce commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,115 @@
+  props: {
+    mailchimp,
+    server: { propDefinition: [mailchimp, "server"] },
+    storeId: {
+      type: "string",
+      label: "Store Id",
+      description:
+        "The unique ID of the store you'd like to watch for new orders. Leave empty to watch for orders on your Mailchimp account.",
+      optional: true,
+    },
+    campaignId: {
+      type: "string",
+      label: "Campaign Id",
+      description:
+        "Watch only for new orders with a specific Campaign Id value.",
+      optional: true,
+    },
```
These should be obtained through an API call, not expect the user to look for them.
The same with `customerId`.

sergioeliot2039 replied:
This has been previously been addressed.

jverce commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,115 @@
+    let mailchimpOrders;
+    let offset = 0;
+    do {
+      mailchimpOrdersInfo = await this.mailchimp.getAllOrders(
+        this.server,
+        this.storeId,
+        1000,
+        offset,
+        this.campaignId,
+        this.outreachId,
+        this.customerId,
+        this.hasOutreach
+      );
```
Retrieving all the orders during each execution does not scale, and will certainly generate too many duplicates and timeouts. We must checkpoint after each emitted event, and based on the API docs I see that `offset` is the only property that will allow us to do it. Let's refactor this similarly to the other suggestions that use a generator to abstract the pagination details, returning an order object and its "offset" (i.e. its position in the list) so that we have a proper starting point the next time the event source is executed.

sergioeliot2039 replied:
This has been addressed.

jverce commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+      default: false,
+      optional: true,
+    },
+    timer: { propDefinition: [mailchimp, "timer"] },
```
The `common-webhook` module is supposed to contain reusable code for http-based event sources. However, it is being used by timer-based event sources. This is counterintuitive and will hurt maintainability. I suggest we refactor the reusable code in a similar way to [the Sendgrid event sources](https://github.com/PipedreamHQ/pipedream/tree/master/components/sendgrid/sources/common)

sergioeliot2039 replied:
This has been previously been addressed.

jverce reviewed (changes requested):
Most of my comments apply to multiple parts of the code, but I only made the comments just to avoid being repetitive.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+  type: "app",
+  app: "mailchimp",
+  propDefinitions: {
+    server: {
+      type: "string",
+      label: "server",
+      description:
+        "The Mailchimp server of the connected account. Found when logging into the Mailchimp account and looking at the URL on a browser. E.g. for https://us19.admin.mailchimp.com/; the us19 part is the server prefix.",
```
ok, i changed the code to get server info from the `$auth` variable.

js07 commented on components/mailchimp/sources/common/http-based.js:
```diff
@@ -0,0 +1,95 @@
+    ...base.hooks,
+    async activate() {
+      const eventsCfg = this.getEventsConfig();
+      const webhook = await this.mailchimp.createWebhook(
```
Source [deployment](https://pipedream.com/docs/cli/reference/#pd-dev) fails because an error is thrown when creating a Mailchimp webhook in the `activate` hook.

sergioeliot2039 replied:
This has been previously addressed.

js07 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,369 @@
+    async createWebhook(listId, webhookUrl, events, sources) {
+      const mailchimp = this._initMailchimpClient();
+      return await this._withRetries(() =>
+        mailchimp.lists.createListWebhook(listId, {
```
Mailchimp responds with the error message:
```
"We couldn't verify the URL is working. Please double check and try again. HTTP Code: 400"
```

sergioeliot2039 replied:
hey @js07 , thanks for this.
i troubleshooted trying to find an issue with the code base from different angles.
in conclusion, i believe this is a problem with Pipedream's Sources new endpoints. Why with new sources? Because if I register a webhook on Mailchimp using the URL from an old endpoint, the webhook gets created correctly. If I use the endpoint that is returned by `$.interface.http` , registering the webook fails. For my testing, I hardcoded in the `http-base.js` file the URL `https://503b39783ae49ef8ae6a07b05f64100b.m.pipedream.net` which I copy pasted from one of my old sources (one I had used for Mailgun), and using this URL works.
That said, the issue might due to slow response to Mailchimp when they try to verify the URL. You can see that I hardcoded the URL in `http-base.js` and this registers properly with Mailchimp.

in details, this is what I did for my troubleshooting:
1. I created an empty source on a local file in my environment, test.js, to confirm I am properly importing attributes from `base.js` and `http-base.js`. in this empty source, I `console.out` variables returned by the $.interface.http, the $.auths variable, etc. just to make sure all things needed are in place
2. I arranged a workflow where I put all the methods used in the sources files, and creating the webhook using this workflow works. I tried adding a local webpage thorugh ngrok, and using the URL provided by ngrok to register, this worked. I also tried with a website of my own, and by the URL from old Sources mentioned before.
3. I added the instruction  `this.http.respond({      status: 200,    });` to acknoledge Mailchimp the eevent. I've seen other developers have used this in the `run` method of the source. This didn't help.

so, can you please have a look again putting special attention to the URL returned by `$.interface.http` and the time it takes to respond?

maybe this is helpful, i created a few workflows, one to return the Mailchimp audiences list, this is a prop for the `new-subscriber` source which I used for testing, a workflow to create a webhook, a workflow to delete a workflow, and a workflow to list workfows. Important, a URL must be used once in a webook list, otherwise Mailchimp will return 400 Bad request.

List audiences
https://pipedream.com/@sergio/mailchimp-list-audience-lists-p_q6CqoKg
Create webhook (based on the sources, http-based code)
https://pipedream.com/@sergio/mailchimp-create-webhook-test-new-subscribe-source-troubleshoot-copy-2-p_95C1LmQ
Delete webhook
https://pipedream.com/@sergio/mailchimp-delete-webhooks-legacy-code-p_3nCbvBL
List webhooks
https://pipedream.com/@sergio/mailchimp-list-webhooks-legacy-code-p_xMCvkGW

js07 replied:
When you added the instruction `this.http.respond({ status: 200, });` to the `run` method, did it also run in cases where `body` was empty. For example:
```js
  async run(event) {
    this.http.respond({
      status: 200,
    });
    const { body } = event;
    if (!body) {
      //Probably a Mailchimp response sent on webhook registration.
      console.log("No body present in event");
      return;
    }
    this.processEvent(body);
  },
```
```js
  async run(event) {
    const { body } = event;
    if (!body) {
      //Probably a Mailchimp response sent on webhook registration.
      this.http.respond({
        status: 200,
      });
      console.log("No body present in event");
      return;
    }
    this.processEvent(body);
  },
```
Just making sure because I think the conditional `!body` will be true when mailchimp is testing the webhook URL on registration.

js07 replied:
The following successfully deploys a Mailchimp source for me:
- Checkout https://github.com/PipedreamHQ/pipedream/pull/1289/commits/86ef7047bc822d0fd4a8da0c8daf91d256252d01
- Change the [`run` method in `common/http-based.js`](https://github.com/PipedreamHQ/pipedream/blob/86ef7047bc822d0fd4a8da0c8daf91d256252d01/components/mailchimp/sources/common/http-based.js#L86-L94) to
```js
  async run(event) {
    this.http.respond({
      status: 200,
    });
    const { body } = event;
    if (!body) {
      //Probably a Mailchimp response sent on webhook registration.
      console.log("No body present in event");
      return;
    }
    this.processEvent(body);
  },
```
- Deploy `new-subscriber.js` using the **Audience List Id** found here:
https://us20.admin.mailchimp.com/lists/members -> Settings -> Audience name and defaults
```
pd dev components/mailchimp/sources/new-subscriber/new-subscriber.js
```

sergioeliot2039 replied:
This has previously been addressed.

js07 commented on components/mailchimp/sources/common/http-based.js:
```diff
@@ -0,0 +1,95 @@
+      return eventsCfg;
+    },
+  },
+  async run(event) {
+    const { body } = event;
+    if (!body) {
+      //Probably a Mailchimp response sent on webhook registration.
+      console.log("No body present in event");
+      return;
+    }
+    this.processEvent(body);
```
On webhook registration, Mailchimp requires a [response](https://mailchimp.com/developer/marketing/guides/sync-audience-data-webhooks/#set-up-your-callback-url) from the webhook endpoint to successfully create the webhook. For example, see [sengrid/sources/common/http-based.js](https://github.com/PipedreamHQ/pipedream/blob/51aea2340b4a4aad158d24166af153597c0399b7/components/sendgrid/sources/common/http-based.js#L75)

```js
  async run(event) {
    this.http.respond({
      status: 200,
    });
```

sergioeliot2039 replied:
Yeah, I tried this but `pd dev` still fails to deploy. I provided details in another comment.

sergioeliot2039 replied:
ok this has previously been addressed and is working properly.

js07 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,369 @@
+     * @params {String} webhookId - The unique ID of the webhook you'd like to delete.
+     * @returns {} - This method returns an empty object.
+     */
+    async deleteWebhook(server, listId, webhookId) {
+      const mailchimp = this._initMailchimpClient(server);
+      return await this._withRetries(() =>
+        mailchimp.lists.deleteListWebhook(listId, webhookId));
```
We'll need to also remove the `server` parameter from the `deleteWebhook` function for the [`deactivate` hook](https://github.com/PipedreamHQ/pipedream/blob/86ef7047bc822d0fd4a8da0c8daf91d256252d01/components/mailchimp/sources/common/http-based.js#L46) to run successfully and allow `pd dev` to update the source.

```suggestion
    async deleteWebhook(listId, webhookId) {
      const mailchimp = this._initMailchimpClient();
      return await this._withRetries(() =>
        mailchimp.lists.deleteListWebhook(listId, webhookId));
```

sergioeliot2039 replied:
Yes, I had this on my backlog.
btw, @dylburger , something I noticed regarding an account's Mailchimp server, is that this is available via the `$.auths` variable when you are developing components. however, when connecting Mailchimp to run node.js (i.e legacy actions), the `auths` variable doesn't include it.

sergioeliot2039 replied:
This previously has been addressed.
A `_server()` method been defined that returns the server from the `$.auths` object.

js07 reviewed (commented):
I made a few comments to address the issue deploying http-based sources as mentioned in [this slack message](https://pipedream-users.slack.com/archives/CPTJYRY5A/p1637268183341000).

sergioeliot2039 commented:
@js07 @dannyroosevelt this is ready for a new review.
main changes:
completed migration to use SDK 
added props definition for reuse across all components
added dynamically populated props where possible
refactored common source logic into base, http and timer based, as recommended elsewhere by Jay
several ESLint fixes.

js07 commented:
@sergioeliot2039 the refactoring into common modules and migration to the MailChimp SDK look good.

Could you resolve the conversations from jverce's review https://github.com/PipedreamHQ/pipedream/pull/1289#pullrequestreview-706842288 above? Many of my comments are the same.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+      const mailchimp = require("@mailchimp/mailchimp_marketing");
+      mailchimp.setConfig({
+        accessToken: this._authToken(),
+        server,
```
This has been implemented.
server property in config object is set via method `this._server()`

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+        try {
+          return await apiCall();
+        } catch (err) {
+          const statusCode = get(err, ["response", "status"]);
```
Jay's suggestion has been included

sergioeliot2039 commented on components/mailchimp/sources/common-webhook.js:
```diff
@@ -0,0 +1,91 @@
+    getEventType() {
+      throw new Error("getEventType is not implemented");
+    },
+    emitEvent(eventPayload) {
```
This has been previously included to the changes.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+          return await apiCall();
+        } catch (err) {
+          const statusCode = get(err, ["response", "status"]);
+          if (!this._isRetriableStatusCode(statusCode)) {
```
I think by the time this review was done the method had not been defined, or maybe I forgot to commit the change back then.
However, it is implemented at this point in time I am commeting.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+          sortDir: "DESC",
+        })
+      );
+    },
```
Great review comment!
Coming from a long C# coding career I was not very used to the opts, config objects I see typical in node.js. 
I am used to them by now, so I have updated the code base to include them.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+        opts.sortField = "create_time";
+      }
+      const mailchimp = this._mailchimp(server);
+      return await this._withRetries(() => mailchimp.campaigns.list(opts));
```
This change has been included.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+     * @params {Integer} offset - For pagination, this the number of records from a collection to skip. Default value is 0.
+     * @params {Date} beforeDate - Restrict response to marketing campaigns created or sent before the set date. Uses ISO 8601 time format: 2015-10-21T15:41:36+00:00.
+     * @params {Date} sinceDate - Restrict response to marketing campaigns created or sent since the set date. Uses ISO 8601 time format: 2015-10-21T15:41:36+00:00.
+     * @returns {campaigns: array, total_items: integer, constraints: object, _links: object } An array with the information of the `campaigns` returned, `total_items` with the total count of the campaigns collection, `constraints` object with possible limitations when quering this endpoint, and `_links` array of links for campaigns manipulation through the Mailchimp API.
```
This as adjusted as in the review.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+    status: {
+      type: "string",
+      label: "status",
+      description: "Campaign status to watch for.",
```
This has been previously added to the changes.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+        console.log("No data available, skipping iteration");
+        return;
+      }
+      const sinceDate = this.mailchimp._statusIsSent(this.status)
+        ? mailchimpCampaigns[0].send_time
+        : mailchimpCampaigns[0].create_time;
```
Eagle eyed catch!!
All right, I have made the refactor shown and then applied the suggested changes whereever in the code applied.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+        null,
+        null
+      );
+      const { campaigns: mailchimpCampaigns = [] } = mailchimpCampaignsInfo;
```
Right, this has beeen changed already.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+      const eventDatedAt = this.mailchimp._statusIsSent(eventPayload.status)
+        ? eventPayload.send_time
+        : eventPayload.create_time;
+      const ts = +new Date(eventDatedAt);
```
This has been included in the changes.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+          beforeDateCreated,
+          sinceDateCreated,
+          sortField: "date_created",
+          sortDir: "DESC",
```
This has been included in the changes.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+  hooks: {
+    async deploy() {
+      // Emits sample events on the first run during deploy.
+      const mailchimpCampaignsInfo = await this.mailchimp.getMailchimpCampaigns(
```
Methods were renamed as suggested.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+      this.db.set("lastSinceDate", sinceDate);
+      mailchimpCampaigns.reverse().forEach(this.emitEvent);
+      offset = offset + mailchimpCampaigns.length;
+    } while (mailchimpCampaigns.length > 0);
```
This suggestion was included in the changes.

sergioeliot2039 commented on components/mailchimp/sources/new-audience/new-audience.js:
```diff
@@ -0,0 +1,79 @@
+      }
+      mailchimpAudienceLists.forEach(this.emitEvent);
+      this.db.set("lastDateCreated", mailchimpAudienceLists[0].date_created);
+      offset = offset + mailchimpAudienceLists.length;
```
This change suggestion has been included-

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+    do {
+      mailchimpCampaignsInfo = await this.mailchimp.getMailchimpCampaigns(
+        this.server,
+        1000,
```
This is sidedly has been addressed by the change related to passing optional parameters on an object.

> Can we instead use opt objects? Having a long list of optional parameters makes the code less readable.
on this [review comment](https://github.com/PipedreamHQ/pipedream/pull/1289#r670914853)

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+        opts.beforeCreateTime = beforeDate;
+        opts.sinceCreateTime = sinceDate;
+        opts.sortField = "create_time";
+      }
```
Nice, this has been implented.

sergioeliot2039 commented on components/mailchimp/sources/common-webhook.js:
```diff
@@ -0,0 +1,91 @@
+    },
+    emitEvent(eventPayload) {
+      const eventTypes = this.getEventType();
+      if (eventTypes.includes(eventPayload.type)) {
```
Interesting, I have applied this suggestion.

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+      sinceDate = this.mailchimp._statusIsSent(this.status)
+        ? mailchimpCampaigns[0].send_time
+        : mailchimpCampaigns[0].create_time;
+      this.db.set("lastSinceDate", sinceDate);
```
these both setting db prop value after events emitted and using getter setters for db prop access have been implemented

sergioeliot2039 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,102 @@
+        ? mailchimpCampaigns[0].send_time
+        : mailchimpCampaigns[0].create_time;
+      this.db.set("lastSinceDate", sinceDate);
+      mailchimpCampaigns.reverse().forEach(this.emitEvent);
```
I implemented this change. However there `.reverse` call had been suggested elsewhere probably in another context with different reviewer. Anyway I understand the changes, records being retrieved then emitted as events look good.

sergioeliot2039 commented on components/mailchimp/sources/new-audience/new-audience.js:
```diff
@@ -0,0 +1,79 @@
+    mailchimp,
+    server: { propDefinition: [mailchimp, "server"] },
+    timer: { propDefinition: [mailchimp, "timer"] },
+    db: "$.service.db",
```
This has been previously addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+      label: "Audience List Id",
+      description:
+        "The unique ID of the audience list related to the segment or tag you'd like to watch for new subscribers.",
+    },
```
This has been previously been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+      label: "Segment/Tag Id",
+      description:
+        "The unique ID of the segment or tag you'd like to watch for new subscribers.",
+    },
```
This has been previously been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+      description:
+        "If set to `true`, it will include subscribers from Mailchimp Marketing and Mailchimp Transactional, formerly Mandrill.  When set to `false`, it will include subscribers from Mailchimp Marketing only.",
+      default: false,
+      optional: true,
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+          sources,
+        })
+      );
+    },
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/common-webhook.js:
```diff
@@ -0,0 +1,91 @@
+      label: "Trigger on subscriber actions?",
+      description:
+        "If set to true, events will be emitted on subscriber-initiated actions.",
+      default: false,
```
Interesting. This has been resolved now.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+          offset,
+        })
+      );
+    },
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+          sinceCreatedAt,
+        })
+      );
+    },
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-file/new-file.js:
```diff
@@ -0,0 +1,89 @@
+      mailchimpFiles.forEach(this.emitEvent);
+      this.db.set("lastCreatedAt", mailchimpFiles[0].created_at);
+      offset = offset + mailchimpFiles.length;
+    } while (mailchimpFiles.length > 0);
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+      if (!mailchimpCampaign) {
+        console.log("No data available, skipping iteration");
+        return;
+      }
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-audience/new-audience.js:
```diff
@@ -0,0 +1,79 @@
+const common = require("../common-webhook");
+const { mailchimp } = common.props;
+const moment = require("moment");
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-audience/new-audience.js:
```diff
@@ -0,0 +1,79 @@
+  methods: {
+    ...common.methods,
+    generateMeta(eventPayload) {
+      const ts = +new Date(eventPayload.created_at);
+      return {
+        id: eventPayload.id,
+        summary: `A new audience list "${eventPayload.name}" was created.`,
+        ts,
+      };
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+
+module.exports = {
+  ...common,
+  key: "mailchimp-new-link",
+  name: "New Link",
```
This has been addressed. The related source file has been renamed accordingly to `link-created.js`

sergioeliot2039 commented on components/mailchimp/sources/new-file/new-file.js:
```diff
@@ -0,0 +1,89 @@
+  hooks: {
+    async deploy() {
+      // Emits sample events on the first run during deploy.
+      const fileType = ["all"].includes(this.fileType) ? null : this.fileType;
```
this has been addressed

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+    emitEvent(eventPayload) {
+      const meta = this.generateMeta(eventPayload);
+      this.$emit(eventPayload, meta);
+    },
```
You are right, I think I first created the overall structure of the component, and hadn't get to abstractions like this one.
This has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+    }
+    const currentRecipientClicks = parseInt(
+      mailchimpCampaign.report_summary.clicks
+    );
```
This has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+      this.$emit(eventPayload, meta);
+    },
+  },
+  async run() {
```
This has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-list-events/new-list-events.js:
```diff
@@ -0,0 +1,54 @@
+        "The unique ID of the audience list you'd like to watch for new subscribers.",
+    },
+    events: {
+      type: "string",
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/common-webhook.js:
```diff
@@ -0,0 +1,91 @@
+    },
+    getEventType() {
+      throw new Error("getEventType is not implemented");
+    },
```
This previously has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-list-events/new-list-events.js:
```diff
@@ -0,0 +1,54 @@
+    },
+    getEventType() {
+      return [this.events];
+    },
```
This has been previously been addressed.

sergioeliot2039 commented on components/mailchimp/sources/common-webhook.js:
```diff
@@ -0,0 +1,91 @@
+      eventsCfg.eventSources.user = this.triggeredByUser;
+      eventsCfg.eventSources.admin = this.triggeredByAdmin;
+      eventsCfg.eventSources.api = this.triggeredByApi;
+      return eventsCfg;
```
This has been previously been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-list-events/new-list-events.js:
```diff
@@ -0,0 +1,54 @@
+module.exports = {
+  ...common,
+  key: "mailchimp-new-list-events",
+  name: "New List Events",
```
This has been been addressed previously.

sergioeliot2039 commented on components/mailchimp/sources/new-list-events/new-list-events.js:
```diff
@@ -0,0 +1,54 @@
+        { label: "Cleaned Email", value: "cleaned" },
+        { label: "Email Address Changes", value: "upemail" },
+        { label: "Campaign sending status", value: "campaign" },
+      ],
```
This has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-list-events/new-list-events.js:
```diff
@@ -0,0 +1,54 @@
+        summary: `New event "${eventPayload.type}" occurred}`,
+        ts,
+      };
+    },
```
This has been implemented.

sergioeliot2039 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,307 @@
+    async getMailchimpCampaignInfo(server, campaignId) {
+      const mailchimp = this._mailchimp(server);
+      return await this._withRetries(() => mailchimp.campaigns.get(campaignId));
+    },
```
Interesting, I have implemented this change.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+      description:
+        "The unique ID of the campaign you'd like to watch for new clicks on links.",
+    },
+    server: { propDefinition: [mailchimp, "server"] },
```
This has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+  },
+  methods: {
+    ...common.methods,
+    generateMeta(eventPayload) {
+      const ts = +new Date();
+      return {
+        id: ts,
+        summary: `Campaign "${eventPayload.settings.title}" now has ${eventPayload.report_summary.clicks} subscriber clicks.`,
+        ts,
+      };
+    },
```
This has been implemented.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+    if (currentRecipientClicks > savedRecipientClicks) {
+      this.db.set("recipientClicks", currentRecipientClicks);
+      this.emitEvent(mailchimpCampaign);
+    }
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-link/new-link.js:
```diff
@@ -0,0 +1,73 @@
+    const mailchimpCampaign = await this.mailchimp.getMailchimpCampaignInfo(
+      this.server,
+      this.campaignId
+    );
```
This has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-file/new-file.js:
```diff
@@ -0,0 +1,89 @@
+    },
+  },
+  async run() {
+    const beforeCreatedAt = moment().toISOString();
```
No, not really. It was just bias to explicitly delimit upper and lower dates, but actually is not needed like you mentioned.

sergioeliot2039 commented on components/mailchimp/sources/new-file/new-file.js:
```diff
@@ -0,0 +1,89 @@
+        return;
+      }
+      this.db.set("lastCreatedAt", mailchimpFiles[0].created_at);
+      mailchimpFiles.forEach(this.emitEvent);
```
this has been addressed.

sergioeliot2039 commented on components/mailchimp/sources/new-open/new-open.js:
```diff
@@ -0,0 +1,74 @@
+      this.emitEvent(mailchimpCampaign);
+    }
+  },
+};
```
Common logic for `new-link` (renamed to `link-opened`) and `new-open` have been abstracted out each source's file modified accordingly.

sergioeliot2039 commented on components/mailchimp/sources/new-customer/new-customer.js:
```diff
@@ -0,0 +1,78 @@
+      }
+      mailchimpStoreCustomers.forEach(this.emitEvent);
+      offset = offset + mailchimpStoreCustomers.length;
+    } while (mailchimpStoreCustomers.length > 0);
```
Very interesting, this has been addressed.

sergioeliot2039 commented:
@js07 @dannyroosevelt this is ready for a new review.

i am required to switch from component development to the app store project. so for the next developer, here is the outstanding:

* most components should be real close if not ready to ship. i expect minor changes on most components except new customer, new order, new open, and link clicked.

* for new customer, and new order, it was suggested that ids of emitted events are tracked, so that they are not emitted again. this is because the related api endpoints, don't have a created/updated  date field that is usually considered to check for new events. it was suggested that all customers or orders are traversed when polling, and each polled object is checked against a list of processed ids. if the id is not found in this list, then the event is emitted and the id is added to the processed ids object. you can see the discussion under [this comment](https://github.com/PipedreamHQ/pipedream/pull/1289#discussion_r6736158064) by @jverce further more i didn't see the need to implement a `_markCustomerAsProcessed` utility function and instead get/set ProcessedIds methods were used. the example referred for a similar strategy used on Twitter is much more complex than the implemented here.

* for new-order and linked-clicked [it was observed by jay](https://github.com/PipedreamHQ/pipedream/pull/1289#discussion_r675245039) that methods from the API reporting module should be used, and since both share common logic from manipulating reports, it was suggested that abstractions are made for said common logic. In my approach i added the `emitReportSampleEvents` and `emitReportEvents` methods under timer-based.js file, then for each file new-order and linked-clicked I added methods that are specific to each source i.e
`cacheCampaignDetails`, and `clearCampaignDetailsCache`; This way the report logic (polling data from the api, calculating the metrics) is done under the same methods across both sources. However, I believe there is space for further abstraction, for instance I see possible to have a new `common-reports` file that the new-open, link-clicked would based from. 

finally, for continuing where I left, it could be a good idea to start with new-audience, new-order and new-customer the reason being is that to test new-audience a mailchimp paid plan must be used, because the free version only allows for 1 audience. for new-order and new-customer, it is needed to connect to a third party store, in my case i used big cartel for the ecommerce store solution... to connect mailchimp and big cartel it was needed to have a paid plan in big cartel. i have subscription for both, i wanted to do this for my testing and make sure the logic implemented was correct. my mailchimp, big cartel subscriptions expire late april, one month from my writing,, if you want that i share credentials, please let me know.

here are the guides i used to connect mailchimp to big cartel store (needed for new order and link clicked)
https://mailchimp.com/help/connect-your-online-store-to-mailchimp/?utm_source=mc-kb&utm_medium=kb-site&utm_campaign=eepurl
https://help.bigcartel.com/mailchimp
since i mentioned these two sources, here are a couple of workflows listing stores, store orders, and list customers, they are public so their related code can be reused
https://pipedream.com/@sergio/mailchimp-list-customers-with-sdk-legacy-code-p_7NCVJND/edit
https://pipedream.com/@sergio/mailchimp-list-store-orders-with-sdk-legacy-code-p_o7CwJgv/edit
https://pipedream.com/@sergio/mailchimp-list-stores-with-sdk-legacy-code-p_7NCVQWY/edit





> We should extract all the common functionality (which is a lot) and only have the event sources specify the report metrics that they want to watch.

michelle0927 commented on .eslintrc.json:
```diff
@@ -0,0 +1,13 @@
+{
+    "env": {
+        "browser": true,
+        "commonjs": true,
+        "es2021": true
+    },
+    "extends": "eslint:recommended",
+    "parserOptions": {
+        "ecmaVersion": 12
+    },
+    "rules": {
+    }
+}
```
Is this change meant to be here?

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+const mailchimp = require("@mailchimp/mailchimp_marketing");
+const retry = require("async-retry");
+module.exports = {
```
I suggest we change `list-event-types.js` to `constants.mjs` and list some additional values there including a default PAGE_SIZE of 1000, since I see that being used in several places.
```suggestion
const retry = require("async-retry");
const constants = require("./sources/constants.mjs");
// or import constants from "./sources/constants.mjs";

module.exports = {
```

```
//constants.mjs

export default {
  PAGE_SIZE: 1000,
  CAMPAIGN_TYPES: [
    "regular",
    "plaintext",
    "absplit",
    "rss",
    "variate",
  ],
  CAMPAIGN_STATUSES: [
    "created",
    "sent",
  ],
  FILE_TYPES: [
    "image",
    "file",
    "all",
  ],
  SEGMENT_WATCH_TYPES: [
    "Created",
    "Updated",
  ],
  ORDER_HAS_OUTREACH_CHOICES: [
    "Yes",
    "No",
    "Both",
  ],
  EVENT_TYPES: [
    {
      label: "Subscribes",
      value: "subscribe",
    },
    {
      label: "Unsubscribes",
      value: "unsubscribe",
    },
    {
      label: "Profile Updates",
      value: "profile",
    },
    {
      label: "Cleaned Email",
      value: "cleaned",
    },
    {
      label: "Email Address Changes",
      value: "upemail",
    },
    {
      label: "Campaign sending status",
      value: "campaign",
    },
  ];
}
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+        "The unique ID of the audience list you'd like to watch for events.",
+      useQuery: true,
+      async options({ page }) {
+        const count = 1000;
+        const offset = 1000 * page;
```
```suggestion
        const count = constants.PAGE_SIZE;
        const offset = count * page;
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+        }));
+      },
+    },
+    // eslint-disable-next-line pipedream/props-description
+    campaignId: {
+      type: "string",
+      label: "Campaign Id",
+      useQuery: true,
```
Remove the commented line and add a description?

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+      label: "Campaign Id",
+      useQuery: true,
+      async options({ page }) {
+        const count = 1000;
+        const offset = 1000 * page;
```
```suggestion
        const count = constants.PAGE_SIZE;
        count offset = count * page;
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+        "If set to `true`, it will include subscribers from Mailchimp Marketing and Mailchimp Transactional, formerly Mandrill.  When set to `false`, it will include subscribers from Mailchimp Marketing only.",
+      default: false,
+    },
+    // eslint-disable-next-line pipedream/props-description
+    storeId: {
+      type: "string",
+      label: "Store Id",
+      useQuery: true,
```
Remove the commented line and add a description?

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+      label: "Store Id",
+      useQuery: true,
+      async options({ page }) {
+        const config = {
+          count: 1000,
+          offset: 1000 * page,
+        };
```
```suggestion
        const count = constants.PAGE_SIZE;
        const config = {
          count,
          offset: count * page,
        };
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+        "The unique ID of the segment or tag you'd like to watch for new subscribers.",
+      useQuery: true,
+      async options(opts) {
+        const count = 1000;
+        const offset = 1000 * opts.page;
+        const config = {
+          count: count,
+          offset: offset,
+        };
```
```suggestion
        const count = constants.PAGE_SIZE;
        const offset = count * opts.page;
        const config = {
          count,
          offset,
        };
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+    statusIsSent(status) {
+      return "sent" === status;
+    },
+    /**
+     * Create a webhook on the specified Audience List.
+     * @param {String} listId - The unique ID that identifies the Audience List you would like to
+     * create a webhook on.
+     * @param {String} config.url - URL listening for the webhook event.
+     * @param {String} config.events -
+     * @param {String} config.sources -
+     * @returns {_links: array, events: object, id: string, list_id: string, sources: object, url:
+     * string} Object with array of links for the webhook manipulation through the Mailchimp API,
+     * an object with flags of the Mailchimp Marketing indicates the events being watched for, an
+     * unique ID of the webhoom, the unique ID which identifies the Audience List where the webhoook
+     * was created, a `sources` object indicating whether the sources for the event are an admin
+     * user, an audience subscriber, or via the API, the URL registered for the webhook event.
+     */
```
I don't think we need to document each value in `config` (url, events, sources). You could document just config and include a link to the relevant documentation. This applies to all methods with `config` as a parameter.

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+    getCampaignsByCreationDate({
+      startDateTime = new Date(0),
+      endDateTime = new Date(),
+      count = 1000,
```
```suggestion
      count = constants.PAGE_SIZE,
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+    getCampaignsBySentDate({
+      startDateTime = new Date(0),
+      endDateTime = new Date(),
+      count = 1000,
```
```suggestion
      count = constants.PAGE_SIZE,
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+     */
+    async *getFileStream(fileType = "all", config) {
+      let offset = 0;
+      const type = fileType === "all"
+        ? null
+        : fileType;
+        const opts = {
+          offset,
+          type,
+          ...config
+        };
+        do {
+          opts.offset = offset;
+          const { files = [] } = await this._withRetries(
+            () => this.api().fileManager.files(opts),
+          );
+          if (files.length === 0) {
+            return;
+          }
+          for (const file of files) {
+            yield file;
+          }
+        offset += files.length;
```
The indentation looks a bit off here.
```suggestion
      const type = fileType === "all"
        ? null
        : fileType;
      const opts = {
        offset,
        type,
        ...config
      };
      do {
        opts.offset = offset;
        const { files = [] } = await this._withRetries(
          () => this.api().fileManager.files(opts),
        );
        if (files.length === 0) {
          return;
        }
        for (const file of files) {
          yield file;
        }
        offset += files.length;
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+      config.customerId = config.customerId === ""
+        ? null
+        : config.customerId;
+      config.hasOutreach = config.hasOutreach === "Both"
+        ? null
+        : config.hasOutreach;
+        do {
+          const opts = {
+            ...config,
+            offset,
+          };
+          const { orders = [] } = await this._withRetries(
+            () => this.api().ecommerce.getStoreOrders(storeId, opts),
+          );
+          if (orders.length === 0) {
+            return;
+          }
+          for (const order of orders) {
+            yield order;
+          }
+        offset += orders.length;
```
The indentation looks a bit off here.
```suggestion
      config.hasOutreach = config.hasOutreach === "Both"
        ? null
        : config.hasOutreach;
      do {
        const opts = {
          ...config,
          offset,
        };
        const { orders = [] } = await this._withRetries(
          () => this.api().ecommerce.getStoreOrders(storeId, opts),
        );
        if (orders.length === 0) {
          return;
        }
        for (const order of orders) {
          yield order;
        }
        offset += orders.length;
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+      {
+        startDateTime = new Date(0),
+        endDateTime = new Date(),
+        count = 1000,
```
```suggestion
        count = constants.PAGE_SIZE,
```

michelle0927 commented on components/mailchimp/mailchimp.app.js:
```diff
@@ -0,0 +1,681 @@
+      {
+        startDateTime = new Date(0),
+        endDateTime = new Date(),
+        count = 1000,
```
```suggestion
        count = constants.PAGE_SIZE,
```

michelle0927 commented on components/mailchimp/sources/common/base.js:
```diff
@@ -0,0 +1,37 @@
+    },
+  },
+  async run(event) {
+    console.log("check about interval times on timer based sources of 60*15 seconds (15 minutes)");
```
Is this console.log meant to be here?

michelle0927 commented on components/mailchimp/sources/common/http-based.js:
```diff
@@ -0,0 +1,114 @@
+  ...base,
+  props: {
+    ...base.props,
+    // eslint-disable-next-line pipedream/props-description, pipedream/props-label
+    http: {
```
I don't think the commented line is needed for `http`.
```suggestion
    http: {
```

michelle0927 commented on components/mailchimp/sources/common/timer-based.js:
```diff
@@ -0,0 +1,66 @@
+  ...base,
+  props: {
+    ...base.props,
+    // eslint-disable-next-line pipedream/props-label, pipedream/props-description
+    timer: {
```
I don't think this commented line is needed for `timer`.
```suggestion
    timer: {
```

michelle0927 commented on components/mailchimp/sources/link-clicked/link-clicked.js:
```diff
@@ -0,0 +1,101 @@
+    },
+  },
+  hooks: {
+    async deploy() {
+      // Emits sample events on the first run during deploy.
+      return await this.emitReportSampleEvents(this.campaignId, this.linkId, (new Date()).getTime());
+    },
```
This `await` can be removed.
```suggestion
    async deploy() {
      // Emits sample events on the first run during deploy.
      return this.emitReportSampleEvents(this.campaignId, this.linkId, (Date.now());
    },
```

michelle0927 commented on components/mailchimp/sources/link-clicked/link-clicked.js:
```diff
@@ -0,0 +1,101 @@
+      return "No data found for specified campaign and link.";
+    },
+    getCurrentCampaignDetails(report) {
+      return  this.uniqueClicksOnly
```
There's an extra space here.
```suggestion
      return this.uniqueClicksOnly
```

michelle0927 commented on components/mailchimp/sources/link-clicked/link-clicked.js:
```diff
@@ -0,0 +1,101 @@
+      ? report.unique_clicks
+      : report.total_clicks;
+    },
+    getDetailsDiff(currentRecipientClicks, recipientClicks) {
+        return currentRecipientClicks - recipientClicks;
+    },
```
The indentation's a little off here.
```suggestion
    getDetailsDiff(currentRecipientClicks, recipientClicks) {
      return currentRecipientClicks - recipientClicks;
    },
```

michelle0927 commented on components/mailchimp/sources/new-audience/new-audience.js:
```diff
@@ -0,0 +1,61 @@
+      // Emits sample events on the first run during deploy.
+      const audiences =
+        await this.mailchimp.getAudienceLists({
+          count: 1000,
```
```suggestion
          count: constants.PAGE_SIZE,
```

michelle0927 commented on components/mailchimp/sources/new-audience/new-audience.js:
```diff
@@ -0,0 +1,61 @@
+  async run() {
+    const sinceDateCreated = this.getDbServiceVariable("lastDateCreated");
+    const beforeDateCreated = (new Date).toISOString();
+    const pageSize = 1000;
```
```suggestion
    const pageSize = constants.PAGE_SIZE;
```

michelle0927 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,101 @@
+const common = require("../common/timer-based");
```
To include `constants.mjs`
```suggestion
const common = require("../common/timer-based");
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,101 @@
+      label: "type",
+      description:
+        "There are four types of campaigns you can create in Mailchimp. A/B Split campaigns have been deprecated and variate campaigns should be used instead. ",
+      options: [
+        "regular",
+        "plaintext",
+        "absplit",
+        "rss",
+        "variate",
+      ],
```
```suggestion
      options: constants.CAMPAIGN_TYPES,
```

michelle0927 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,101 @@
+      type: "string",
+      label: "status",
+      description: "Campaign status to watch for.",
+      options: [
+        "created",
+        "sent",
+      ],
```
```suggestion
      options: constants.CAMPAIGN_STATUSES,
```

michelle0927 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,101 @@
+        count: 10,
+        offset: 0,
+      };
+      let campaigns;
+      if(this.mailchimp.statusIsSent(this.status)) {
+        campaigns = await this.mailchimp.getCampaignsBySentDate(config);
+      } else {
+        campaigns = await this.mailchimp.getCampaignsByCreationDate(config);
+      }
```
Could simplify it this way:
```suggestion
      const campaigns = this.mailchimp.statusIsSent(this.status) 
        ? await this.mailchimp.getCampaignsBySentDate(config) 
        : await this.mailchimp.getCampaignsByCreationDate(config);
```

michelle0927 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,101 @@
+  },
+  async run() {
+    const beforeDate = (new Date).toISOString();
+    const pageSize = 1000;
```
```suggestion
    const pageSize = constants.PAGE_SIZE;
```

michelle0927 commented on components/mailchimp/sources/new-campaign/new-campaign.js:
```diff
@@ -0,0 +1,101 @@
+    do {
+      config.sinceDate = sinceDate;
+      config.offset = offset;
+      if(this.mailchimp.statusIsSent(this.status)) {
+        campaigns = await this.mailchimp.getCampaignsBySentDate(config);
+      } else {
+        campaigns = await this.mailchimp.getCampaignsByCreationDate(config);
+      }
```
```suggestion
      campaigns = this.mailchimp.statusIsSent(this.status)
        ? await this.mailchimp.getCampaignsBySentDate(config)
        : await this.mailchimp.getCampaignsByCreationDate(config);
```

michelle0927 commented on components/mailchimp/sources/new-customer/new-customer.js:
```diff
@@ -0,0 +1,74 @@
+const common = require("../common/timer-based");
```
To include `constants.mjs`
```suggestion
const common = require("../common/timer-based");
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-customer/new-customer.js:
```diff
@@ -0,0 +1,74 @@
+  },
+  async run() {
+    const processedIds = new Set(this._getProcessedIds());
+    const pageSize = 1000;
```
```suggestion
    const pageSize = constants.PAGE_SIZE;
```

michelle0927 commented on components/mailchimp/sources/new-file/new-file.js:
```diff
@@ -0,0 +1,68 @@
+const common = require("../common/timer-based");
```
To include `constants.mjs`
```suggestion
const common = require("../common/timer-based");
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-file/new-file.js:
```diff
@@ -0,0 +1,68 @@
+      label: "File type",
+      description:
+        "The file type you'd like to watch for in the File Manager. Supported file types: 'image' or 'file', use 'all' for both.",
+      options: [
+        "image",
+        "file",
+        "all",
+      ],
```
```suggestion
      options: constants.FILE_TYPES,
```

michelle0927 commented on components/mailchimp/sources/new-list-event/new-list-event.js:
```diff
@@ -0,0 +1,42 @@
+const common = require("../common/http-based");
+const listEventTypes = require('../list-event-types.js');
```
If you change `list-event-types` to `constants`
```suggestion
const common = require("../common/http-based");
//const listEventTypes = require('../list-event-types.js');
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-list-event/new-list-event.js:
```diff
@@ -0,0 +1,42 @@
+      description:
+        "The events to be emitted, which will trigger the source being created.",
+      default: ["subscribe"],
+      options: listEventTypes,
```
```suggestion
      options: constants.EVENT_TYPES,
```

michelle0927 commented on components/mailchimp/sources/new-open/new-open.js:
```diff
@@ -0,0 +1,97 @@
+    },
+  },
+  hooks: {
+    async deploy() {
+      // Emits sample events on the first run during deploy.
+      return await this.emitReportSampleEvents(this.campaignId, undefined, (new Date()).getTime());
```
This `await` can be removed
```suggestion
    async deploy() {
      // Emits sample events on the first run during deploy.
      return this.emitReportSampleEvents(this.campaignId, undefined, (Date.now());
```

michelle0927 commented on components/mailchimp/sources/new-open/new-open.js:
```diff
@@ -0,0 +1,97 @@
+        members: this.getDbServiceVariable("members"),
+      };
+    },
+    async getCampaignDetails(){
+      return await this.mailchimp.listCampaignOpenDetails(
+        this.campaignId,
+      );
```
This `await` can be removed
```suggestion
    async getCampaignDetails(){
      return this.mailchimp.listCampaignOpenDetails(
        this.campaignId,
      );
```

michelle0927 commented on components/mailchimp/sources/new-open/new-open.js:
```diff
@@ -0,0 +1,97 @@
+      }
+    },
+    getDetailsDiff(currentCampaignOpenReport, cachedCampaignOpenReport) {
+      if(this.uniqueOpensOnly) {
+        return (currentCampaignOpenReport.members.length | 0) - (cachedCampaignOpenReport.members.length | 0);
+      } else {
+        return currentCampaignOpenReport.total_opens - cachedCampaignOpenReport.total_opens;
+      }
```
Could simplify it this way:
```suggestion
      if(this.uniqueOpensOnly) {
        return (currentCampaignOpenReport.members.length | 0) - (cachedCampaignOpenReport.members.length | 0);
      }
      return currentCampaignOpenReport.total_opens - cachedCampaignOpenReport.total_opens;
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,110 @@
+const common = require("../common/timer-based");
```
To include `constants.mjs`
```suggestion
const common = require("../common/timer-based");
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,110 @@
+      label: "Watch for new created or updated segments?",
+      description:
+        "If set to `Created`, it will include new created segments only.  When set to `Updated`, it will only include recently updated segments.",
+      options: [
+        "Created",
+        "Updated",
+      ],
```
```suggestion
      options: constants.SEGMENT_WATCH_TYPES,
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,110 @@
+        count: 10,
+        offset: 0,
+      };
+     let segments;
+     if(this.watchFor === "Created" ) {
+      segments = await this.mailchimp.getAudienceSegmentsByCreatedDate(this.listId, config);
+     } else {
+      segments = await this.mailchimp.getAudienceSegmentsByUpdatedDate(this.listId, config);
+     }
```
Could simplify it this way:
```suggestion
     const segments = this.watchFor === "Created"
        ? await this.mailchimp.getAudienceSegmentsByCreatedDate(this.listId, config)
        : segments = await this.mailchimp.getAudienceSegmentsByUpdatedDate(this.listId, config);
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,110 @@
+    const endDateTime = (new Date).toISOString();
+    let segments;
+    let offset = 0;
+    const pageSize = 1000;
```
```suggestion
    const pageSize = constants.PAGE_SIZE;
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js:
```diff
@@ -0,0 +1,110 @@
+    do {
+      config.offset = offset;
+      config.startDateTime = startDateTime;
+      if(this.watchFor === "Created" ) {
+       segments = await this.mailchimp.getAudienceSegmentsByCreatedDate(this.listId, config);
+      } else {
+       segments = await this.mailchimp.getAudienceSegmentsByUpdatedDate(this.listId, config);
+      }
```
Could simplify it this way:
```suggestion
      segments = this.watchFor === "Created"
        ? segments = await this.mailchimp.getAudienceSegmentsByCreatedDate(this.listId, config)
        : segments = await this.mailchimp.getAudienceSegmentsByUpdatedDate(this.listId, config);
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-subscriber/new-or-updated-subscriber.js:
```diff
@@ -0,0 +1,42 @@
+module.exports = {
+  ...common,
+  key: "new-or-updated-subscriber",
+  name: "New Or Updated Subscriber",
```
```suggestion
  name: "New or Updated Subscriber",
```

michelle0927 commented on components/mailchimp/sources/new-or-updated-subscriber/new-or-updated-subscriber.js:
```diff
@@ -0,0 +1,42 @@
+    },
+    generateMeta(eventPayload) {
+      const ts = Date.parse(eventPayload.fired_at);
+      let summary;
+      if (this.getEventTypes().includes(eventPayload.type)) {
+        summary =`New event "${eventPayload.type}" occurred`;
+      } else {
+        console.log("Unexpected trigger found, skipping emit");
+        return;
+      }
```
Could we write it this way?
```suggestion
      if (!this.getEventTypes().includes(eventPayload.type)) {
        console.log("Unexpected trigger found, skipping emit");
        return;
      }
      const summary =`New event "${eventPayload.type}" occurred`;
      return {
        id: eventPayload["data[id]"],
        summary,
        ts,
      };
```

michelle0927 commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,151 @@
+const common = require("../common/timer-based");
```
To include `constants.mjs`
```suggestion
const common = require("../common/timer-based");
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,151 @@
+      optional: true,
+      useQuery: true,
+      async options({ page }) {
+        const count = 1000;
+        const offset = 1000 * page;
```
```suggestion
        const count = constants.PAGE_SIZE;
        const offset = count * page;
```

michelle0927 commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,151 @@
+      label: "Has Outreach?",
+      description:
+        "Watch only for new orders that have an outreach attached. For example, an email campaign or Facebook ad.",
+      options: [
+        "Yes",
+        "No",
+        "Both",
+      ],
```
```suggestion
      options: constants.ORDER_HAS_OUTREACH_CHOICES,
```

michelle0927 commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,151 @@
+      optional: true,
+      useQuery: true,
+      async options({ page }) {
+        const count = 1000;
+        const offset = 1000 * page;
```
```suggestion
        const count = constants.PAGE_SIZE;
        const offset = count * page;
```

michelle0927 commented on components/mailchimp/sources/new-order/new-order.js:
```diff
@@ -0,0 +1,151 @@
+  async run() {
+    const processedIds = new Set(this._getProcessedIds());
+    const config = {
+      count: 1000,
```
```suggestion
      count: constants.PAGE_SIZE,
```

michelle0927 commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+const common = require("../common/timer-based");
```
To include `constants.mjs`
```suggestion
const common = require("../common/timer-based");
const constants = require("../constants.mjs");
// or import constants from "../constants.mjs";
```

michelle0927 commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+    },
+  },
+  async run() {
+    const pageSize = 1000;
```
```suggestion
    const pageSize = constants.PAGE_SIZE;
```

michelle0927 commented on components/mailchimp/sources/new-segment-tag-subscriber/new-segment-tag-subscriber.js:
```diff
@@ -0,0 +1,97 @@
+    let mailchimpSegmentMembersInfo;
+    let mailchimpSegmentMembers;
+    let offset = 0;
+    const config = {
+      count: 1000,
+      includeTransactional: this.includeTransactional,
+      includeUnsubscribed: false,
+    };
```
```suggestion
    const config = {
      count: pageSize,
      includeTransactional: this.includeTransactional,
      includeUnsubscribed: false,
    };
```

michelle0927 commented on components/mailchimp/sources/new-subscriber/new-subscriber.js:
```diff
@@ -0,0 +1,30 @@
+  version: "0.0.1",
+  type: "source",
+  dedupe: "unique",
+  props: {
+    ...common.props,
+  },
```
Since no props are being overwritten here, this `props` section can be removed.
```suggestion
```

michelle0927 commented on components/mailchimp/sources/new-unsubscriber/new-unsubscriber.js:
```diff
@@ -0,0 +1,31 @@
+  version: "0.0.1",
+  type: "source",
+  dedupe: "unique",
+  props: {
+    ...common.props,
+  },
```
```suggestion
```

michelle0927 reviewed (changes requested):
Would you be able to update the .js files to .mjs files?

These files are outside the components directory: `eslintrc.json`, `package-lock.json`, and `package.json`. Are they meant to be included in this PR?

There are no `package.json` and `package-lock.json` files yet in the `components/mailchimp` directory, so those should be added.

michelle0927 commented:
One more thing! Remember to run `npx eslint` on everything. Looks like this is failing the Lint check in several places.
